lib/trento/users.ex

```diff
@@ -0,0 +1,78 @@
+defmodule Trento.Users do
+  @moduledoc """
+  The Users context.
+  """
+
+  use Pow.Ecto.Context,
+    repo: Trento.Repo,
+    user: Trento.Users.User
+
+  import Ecto.Query, warn: false
+  alias Trento.Repo
+
+  alias Trento.Users.User
+
+  @impl true
+  @doc """
+    get_by function overrides the one defined in Pow.Ecto.Context,
+    we retrieve the user by username as traditional Pow flow but we also exclude
+    deleted and locked users
+  """
+  def get_by(clauses) do
+    username = clauses[:username]
+
+    User
+    |> where([u], is_nil(u.deleted_at) and is_nil(u.locked_at) and u.username == ^username)
+    |> Repo.one()
+  end
+
+  def list_users do
+    User
+    |> where([u], is_nil(u.deleted_at))
+    |> Repo.all()
+  end
+
+  def get_user(id) do
+    case User
+         |> where([u], is_nil(u.deleted_at) and u.id == ^id)
+         |> Repo.one() do
+      nil -> {:error, :not_found}
+      user -> {:ok, user}
+    end
+  end
+
+  def create_user(attrs \\ %{}) do
+    %User{}
+    |> User.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  def update_user(%User{id: 1}, _), do: {:error, :operation_not_permitted}
+
+  def update_user(%User{locked_at: nil} = user, %{enabled: false} = attrs) do
+    do_update(user, Map.put(attrs, :locked_at, DateTime.utc_now()))
+  end
+
+  def update_user(%User{locked_at: locked_at} = user, %{enabled: false} = attrs)
+      when not is_nil(locked_at) do
```

Comment:
You don't need this `is_nil`. With the pattern matching on top of this, you are assured that it never will be

Reply:
We need this clause to match the case, when the user is already locked, but from the api comes to "relock" again, to not update the locked_at timestamp, it's an edge case, if I remove that clause the tests does not pass for that edge case

Reply:
Ok ok I see. I was not thinking about the lock a locked user hehe